src/lib/misc/roughtime/roughtime.cpp

```diff
@@ -0,0 +1,413 @@
+   auto level = 0u;
+   while(level < levels)
+      {
+      hash = index&1
+             ? hashNode(typecast_copy<std::array<uint8_t, 64>>(path.data() + level*64), hash)
+             : hashNode(hash, typecast_copy<std::array<uint8_t, 64>>(path.data() + level*64));
+      ++level;
+      index>>=1;
+      }
+
+   if(srep_root != hash)
+      { throw Invalid_Argument("nonce verification failed"); }
+
+   const auto cert_dele_maxt = sys_microseconds64(get<microseconds64>(cert_dele_v, "MAXT"));
+   const auto cert_dele_mint = sys_microseconds64(get<microseconds64>(cert_dele_v, "MINT"));
+   const auto srep_midp = sys_microseconds64(get<microseconds64>(srep_v, "MIDP"));
+   const auto srep_radi = get<microseconds32>(srep_v, "RADI");
+   if(srep_midp < cert_dele_mint)
+      { throw Invalid_Argument("midpoint earlier than delegation start"); }
+   if(srep_midp > cert_dele_maxt)
+      { throw Invalid_Argument("midpoint later than delegation end"); }
+   return {cert_dele, cert_sig, srep_midp, srep_radi};
+   }
+
+bool Response::validate(const Ed25519_PublicKey& pk) const
+   {
+   const char context[] = "RoughTime v1 delegation signature--";
+   PK_Verifier verifier(pk, "Pure");
+   verifier.update(cast_char_ptr_to_uint8(context), sizeof(context)); //add context including \0
+   verifier.update(m_cert_dele.data(), m_cert_dele.size());
+   return verifier.check_signature(m_cert_sig.data(), m_cert_sig.size());
+   }
+
+Nonce nonce_from_blind(const std::vector<uint8_t>& previous_response,
+                       const Nonce& blind)
+   {
+   std::array<uint8_t, 64> ret;
+   const auto blind_arr = blind.get_nonce();
+   std::unique_ptr<Botan::HashFunction> hash(Botan::HashFunction::create("SHA-512"));
+   hash->update(previous_response);
+   hash->update(hash->final());
+   hash->update(blind_arr.data(), blind_arr.size());
+   hash->final(ret.data());
+
+   return ret;
+   }
+
+Chain::Chain(const std::string& str)
+   {
+   std::stringstream ss(str);
+   for(std::string s; std::getline(ss, s);)
+      {
+      size_t start = 0, end = 0;
+      end = s.find(' ', start);
+      if(end == std::string::npos)
+         {
+         throw Invalid_Argument("wrong file format");
+         }
+      const auto publicKeyType = s.substr(start, end-start);
+      if(publicKeyType != "ed25519")
+         { throw Not_Implemented("only ed25519 publicKeyType is implemented"); }
+
+      start = end + 1;
+      end = s.find(' ', start);
+      if(end == std::string::npos)
+         {
+         throw Invalid_Argument("wrong file format");
+         }
+      const auto serverPublicKey = Botan::Ed25519_PublicKey(Botan::base64_decode(s.substr(start, end-start)));
+
+      start = end + 1;
+      end = s.find(' ', start);
+      if(end == std::string::npos)
+         {
+         throw Invalid_Argument("wrong file format");
+         }
+      const auto nonceOrBlind = Nonce(vector_to_array<64>(Botan::unlock(Botan::base64_decode(s.substr(start, end-start)))));
```

Comment:
88 bytes

Reply:
@randombit I don't think this is valid. base_decode requires the output array to be larger than the  final length. While base64_decode returns 64, it actually wrote over the 64 buts. The sanitizer caught this for me. This appears to be the intended implementation, see base_decode:
`size_t written = (out_ptr - output) - base.bytes_to_remove(final_truncate);`
Do you have anothe suggestion?

Reply:
> While base64_decode returns 64, it actually wrote over the 64 boundary.

That seems like a bug. I will investigate.

Reply:
> > While base64_decode returns 64, it actually wrote over the 64 boundary.
> 
> That seems like a bug. I will investigate.

@randombit please don't forget this.

Reply:
Thanks I had forgotten. I wrote a fuzzer trying to find the issue but it has found nothing so far. Was the issue reliably reproducible for you?

Reply:
Yes, still the fail you can see I'm CI. I don't understand what are you looking for. See the code I copied above. The part of "base.bytes_to_remove(final_truncate)". This function is designed to need a array with some extra bytes which are trimmed at the end.

Reply:
Sorry I didn't realize the CI failure was related here.

I don't think it is a bug, exactly, though certainly points to a bad API.

When decoding, the length of the output buffer is not specified. Instead it is assumed to always be at least base64_decode_max_output bytes, which is asserted (in a sense) by calling clear_mem on the output buffer zeroing that many bytes. Because the value passed to base64_decode is 88 bytes, with 2 bytes of `=` padding, it decodes to 64 bytes. But base64_decode_max_output doesn't look at contents and only length, so it assumes you have passed 88 bytes of base64 with no padding or whitespace. In that case the output would be 66 bytes, so that is the amount that is memset to zero at the beginning of decoding. But of course 64 is returned since that is the value actually decoded.

Short term fix is to check that the last 2 bytes are == then remove them before doing base64 decoding, which will produce exactly the 64 byte value expected.

It would be best to add some new base64 functions which take an explicit length for the output buffer, which would have found this issue immediately. This length assumption in the API is definitely a nice footgun.

Reply:
I agree it's not a big. The function is just not intended to do what you asked me to use it for. I can though still avoid one copy by copying from secure_vector to array directly. 
I will not be changing base64 code, I don't have time to get this ready this week. And I really want to close this.
Are you doing some work on this or should I go ahead and go with the intermediate solution of memcpy secure vector to array?

Reply:
I also would like to see this merged in time for 2.12 it has been long enough already.

Just copy it for now. I will do the base64 changes + update Roughtime to use them early in 2.13 cycle.